lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java

```diff
@@ -337,9 +354,79 @@ public MavenPluginManager getMavenPluginManager() {
 	}
 
 	public void didChangeWorkspaceFolders(URI[] added, URI[] removed) {
+		initialize();
 		WorkspaceReader workspaceReader = mavenRequest.getWorkspaceReader();
 		if (workspaceReader instanceof MavenLemminxWorkspaceReader) {
-			// TODO
+			Collection<DOMDocument> projectsToAdd = computeAddedWorkspaceProjects(added != null? added : new URI[0]);
+			Collection<URI> projectsToRemove = computeRemovedWorkspaceProjects(removed != null ? removed : new URI[0]);
+
+			if (projectsToAdd != null) {
+				projectsToAdd.stream().forEach(d -> getProjectCache().addDocument(URI.create(d.getDocumentURI()), d));
+			}
+
+			if (projectsToRemove != null) {
+				projectsToRemove.stream().forEach(u -> getProjectCache().removeDocument(u));
+			}
 		}
 	}
+	
+	private Collection<URI> computeRemovedWorkspaceProjects(URI[] removed) {
+		Collection<MavenProject> cachedProjects = getProjectCache().getProjects();
+		return cachedProjects.stream().filter(p -> {
+			return Arrays.asList(removed).stream().filter(uri -> {
+				Path removedPath = new File(uri).toPath();
+				return p.getFile().toPath().startsWith(removedPath);
+			}).findAny().isPresent();
+		}).map(p -> p.getFile().toURI()).collect(Collectors.toUnmodifiableList());
+	}
+	
+	private List<DOMDocument> computeAddedWorkspaceProjects(URI[] added) {
+		List<DOMDocument> projectsToAdd = new ArrayList<DOMDocument>();
+		Arrays.asList(added).stream().forEach(uri -> {
+			Path addedPath = new File(uri).toPath();
+			try {
+				Files.walkFileTree(addedPath, Collections.emptySet(), 10, new SimpleFileVisitor<Path>() {
+					@Override
+					public FileVisitResult preVisitDirectory(Path file, BasicFileAttributes attrs) throws IOException {
+						if (file.getFileName().toString().charAt(0) == '.') {
+							return FileVisitResult.SKIP_SUBTREE;
+						}
+						return FileVisitResult.CONTINUE;
+					}
+
+					@Override
+					public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+						if (file.getFileName().toString().toLowerCase().endsWith(".xml")) {

```

Comment:
This should reuse the same logic as `match` method. Maybe having a `match(File)` method would be good here.

Reply:
I would change it to `match (Path)` as visitFile() operates with Path